policy-controller/k8s/index/src/outbound_index.rs

```diff
@@ -0,0 +1,281 @@
+                }
+            }
+        }
+    }
+
+    fn delete(&mut self, namespace: String, name: String) {
+        let service_ref = ServiceRef { name, namespace };
+        self.services.retain(|_, v| *v != service_ref);
+    }
+}
+
+impl Index {
+    pub fn shared(cluster_domain: String) -> SharedIndex {
+        Arc::new(RwLock::new(Self {
+            namespaces: NamespaceIndex {
+                by_ns: HashMap::default(),
+                cluster_domain: Arc::new(cluster_domain),
+            },
+            services: HashMap::default(),
+        }))
+    }
+
+    pub fn outbound_policy_rx(
+        &mut self,
+        namespace: &str,
+        service: &str,
+        port: NonZeroU16,
+    ) -> Result<watch::Receiver<OutboundPolicy>> {
+        let ns = self
+            .namespaces
+            .by_ns
+            .entry(namespace.to_string())
+            .or_insert_with(|| Namespace {
+                services: Default::default(),
+                namespace: Arc::new(namespace.to_string()),
+                cluster_domain: self.namespaces.cluster_domain.clone(),
+            });
+        let key = ServicePort {
+            service: service.to_string(),
+            port,
+        };
+        tracing::debug!(?key, "subscribing to service port");
+        let routes = ns.service_routes_or_default(key);
+        Ok(routes.watch.subscribe())
+    }
+
+    pub fn lookup_service(&self, addr: IpAddr) -> Option<ServiceRef> {
+        self.services.get(&addr).cloned()
+    }
+}
+
+impl Namespace {
+    fn apply(&mut self, route: HttpRoute) {
+        tracing::debug!(?route);
+        for parent_ref in route.spec.inner.parent_refs.iter().flatten() {
+            if parent_ref.kind.as_deref() == Some("Service") {
+                if let Some(port) = parent_ref.port {
+                    if let Some(port) = NonZeroU16::new(port) {
+                        let service_port = ServicePort {
+                            port,
+                            service: parent_ref.name.clone(),
+                        };
+                        tracing::debug!(
+                            ?service_port,
+                            route = route.name_unchecked(),
+                            "inserting route for service"
+                        );
+                        let service_routes = self.service_routes_or_default(service_port);
+                        service_routes.apply(route.clone());
+                    } else {
+                        tracing::warn!(?parent_ref, "ignoring parent_ref with port 0");
+                    }
+                } else {
+                    tracing::warn!(?parent_ref, "ignoring parent_ref without port");
+                }
+            } else {
+                tracing::warn!(
+                    parent_kind = parent_ref.kind.as_deref(),
+                    "ignoring parent_ref with non-Service kind"
+                );
+            }
```

Comment:
we could potentially reduce rightward drift here by using the new [`let .. else` syntax from Rust 1.65](https://releases.rs/docs/released/1.65.0/#language):
```suggestion
            if parent_ref.kind.as_deref() != Some("Service") {
                 tracing::warn!(
                    parent_kind = parent_ref.kind.as_deref(),
                    "ignoring parent_ref with non-Service kind"
                );
                continue;
            }

            let Some(port) = parent_ref.port else {
                tracing::warn!(?parent_ref, "ignoring parent_ref without port");
                continue;
            };

            let Some(port) = NonZeroU16::new(port) else {
                tracing::warn!(?parent_ref, "ignoring parent_ref with port 0");
                continue;
            };

            let service_port = ServicePort {
                port,
                service: parent_ref.name.clone(),
            };

            tracing::debug!(
                ?service_port,
                route = route.name_unchecked(),
                "inserting route for service"
            );

            let service_routes = self.service_routes_or_default(service_port);
            service_routes.apply(route.clone());
```
personally, i think this might be easier to follow, but your mileage may vary :woman_shrugging: